aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,359 @@
+    /// 2. the y values must be non-decreasing and between (0, 10000);
+    /// So the curve will be identified as point (0, 0) and (10000, 10000) interpolated with the points. The y value
+    /// between two points will be calculated by neighboring points as if there is a linear line connecting these two
+    /// points.
+    struct GasCurve has copy, drop, store {
+        min_gas: u64,
+        max_gas: u64,
+        points: vector<Point>,
+    }
+
+    /// P(x) = min_price + (base ^ (utilization / target_usage) - 1) / (base - 1) * (max_price - min_price)
+    // Provide a default exponential curve with the base to be 8192, which means:
+    // When DB is at 50% target usage,the price increases roughly 1.09% of (max_price - min_price).
+    // Detailed data points:
+    // 10% -> 0.02%
+    // 20% -> 0.06%
+    // 30% -> 0.17%
+    // 40% -> 0.44%
+    // 50% -> 1.09%
+    // 60% -> 2.71%
+    // 70% -> 6.69%
+    // 80% -> 16.48%
+    // 90% -> 40.61%
+    // 95% -> 63.72%
+    // 99% -> 91.38%
+    public fun base_8192_exponential_curve(min_gas: u64, max_gas: u64): GasCurve {
+        new_gas_curve(min_gas, max_gas,
+            vector[
+                new_point(1000, 2),
+                new_point(2000, 6),
+                new_point(3000, 17),
+                new_point(4000, 44),
+                new_point(5000, 109),
+                new_point(6000, 271),
+                new_point(7000, 669),
+                new_point(8000, 1648),
+                new_point(9000, 4061),
+                new_point(9500, 6372),
+                new_point(9900, 9138),
+            ]
+        )
+    }
+
+    struct StorageGasConfig has copy, drop, key {
+        item_config: UsageGasConfig,
+        byte_config: UsageGasConfig,
+    }
+
+    public fun new_point(x: u64, y: u64): Point {
+        Point {x, y}
+    }
+
+    public fun new_gas_curve(min_gas: u64, max_gas: u64, points: vector<Point>): GasCurve {
+        GasCurve {
+            min_gas,
+            max_gas,
+            points
+        }
+    }
+
+    public fun new_usage_gas_config(target_usage: u64, read_curve: GasCurve, create_curve: GasCurve, write_curve: GasCurve): UsageGasConfig {
+        UsageGasConfig {
+            target_usage,
+            read_curve,
+            create_curve,
+            write_curve,
+        }
+    }
+
+    public fun new_storage_gas_config(item_config: UsageGasConfig, byte_config: UsageGasConfig): StorageGasConfig {
+        StorageGasConfig {
+            item_config,
+            byte_config
+        }
+    }
+
+    public(friend) fun set_config(aptos_framework: &signer, config: StorageGasConfig) acquires StorageGasConfig {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        validate_usage_config(&config.item_config);
+        validate_usage_config(&config.byte_config);
+        *borrow_global_mut<StorageGasConfig>(@aptos_framework) = config;
+    }
+
+    fun init_module(aptos_framework: &signer) {
```

Comment:
Did you verify that this works in the context of a package upgrade?

Reply:
Is there an easy way to do that?
is that something `fn init_module()` tries to do?